[flang][debug] Support complex types.

abidh opened the pull request:
This PR adds supports for conversion of complex type to corresponding DITypeAttr. Both fir and mlir types are supported.

Apart from lit testing, I have also tested the types in debugger and they work correctly. An exception is 128 bit complex which somehow requires that its name be different from `complex`. I am going to open a separate PR to add (kind=n) in the type names similar to what gfortran does.

abidh committed:
[flang][debug] Support complex types.

This PR adds supports for conversion of complex type to corresponding
DITypeAttr. Both fir and mlir types are supported.

Apart from lit testing, I have also tested the types in debugger and
they work correctly. An exception is 128 bit complex which somehow
requires that its name be different from `complex`. I am going to open
a separate PR to add (kind=n) in the type names similar to what
gfortran does.

abidh committed:
Move LLVM IR tests to integration folder. (NFC)

tblah commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -57,6 +57,16 @@ DebugTypeGenerator::convertType(mlir::Type Ty, mlir::LLVM::DIFileAttr fileAttr,
+    if (auto cplxTy = mlir::dyn_cast_or_null<mlir::ComplexType>(Ty)) {
+      auto floatTy = mlir::cast<mlir::FloatType>(cplxTy.getElementType());
+      bitWidth = floatTy.getWidth();
+    } else if (auto cplxTy = mlir::dyn_cast_or_null<fir::ComplexType>(Ty))
```
nit: I think this would be clearer to read if you used braces on the else if branch to match the if branch. There is a similar example advocating braces here https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

abidh replied:
Thanks for pointing that out. I will fix it.

tblah commented on flang/test/Transforms/debug-complex-2.f90:
Could this be merged with `debug-complex-1.f90`? Maybe just use different FileCheck prefixes with and without `preserve-declare`? I think that would make it clearer what is different between each test

abidh replied:
They were in the same file when I opened the PR. I just separated them out in commit 9725619e4d3523a304d431a8b248a04a7e33bed8 as @kiranchandramohan pointed out in https://github.com/llvm/llvm-project/pull/91582#discussion_r1606046605 that LLVM IR tests should go in integration folder.

abidh committed:
Handle review comments.

tblah reviewed (approved):
Thanks for this!

kiranchandramohan commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -57,6 +57,16 @@ DebugTypeGenerator::convertType(mlir::Type Ty, mlir::LLVM::DIFileAttr fileAttr,
                         mlir::StringAttr::get(context, logTy.getMnemonic()),
                         kindMapping.getLogicalBitsize(logTy.getFKind()),
                         llvm::dwarf::DW_ATE_boolean);
+  } else if (fir::isa_complex(Ty)) {
```
LLVM has some guidelines regarding else after return. https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

kiranchandramohan commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -57,6 +57,16 @@ DebugTypeGenerator::convertType(mlir::Type Ty, mlir::LLVM::DIFileAttr fileAttr,
                         kindMapping.getLogicalBitsize(logTy.getFKind()),
                         llvm::dwarf::DW_ATE_boolean);
+  } else if (fir::isa_complex(Ty)) {
+    unsigned bitWidth;
```
Would it make sense to initialize this to some value to avoid an uninitialized warning or alternatively add an else case below to assert or llvm unreachable?

kiranchandramohan commented on flang/test/Transforms/debug-complex-1.fir:
```diff
@@ -0,0 +1,39 @@
+// RUN: fir-opt --cg-rewrite="preserve-declare=true" --mlir-print-debuginfo %s | fir-opt --add-debug-info --mlir-print-debuginfo | FileCheck %s
```
Does the cg-rewrite pass do anything useful here?

abidh replied:
No, only add-debug-info is needed. I have updated the test.

kiranchandramohan commented on flang/test/Transforms/debug-complex-2.f90:
```diff
@@ -0,0 +1,28 @@
+! RUN: %flang_fc1 -emit-fir -debug-info-kind=standalone -mmlir --mlir-print-debuginfo %s -o - | \
```
Does this test anything different from `debug-complex-1.fir`?

abidh replied:
I was going to remove it after the discussion in #91582. Done now.

abidh committed:
Handle review comments

1. Remove the fortran testcase as it was redundant.
2. Remove redundant command from RUN line.
3. Add an llvm_unreachable if we get a complex type which is not
   fir or mlir complex type.

kiranchandramohan reviewed (approved):
LG.
